pytorch_lightning/core/hooks.py

```diff
@@ -138,11 +138,20 @@ def backward(self, use_amp, loss, optimizer):
                 else:
                     loss.backward()
 
+        .. note:: with PyTorch 1.6+ + precision=16 + multiple optimizers, set .backward(retrain_graph=True)
+
         """
         if trainer.precision == 16:
-
             # .backward is not special on 16-bit with TPUs
-            if not trainer.on_tpu:
+            if trainer.on_tpu:
+                return
+
+            if self.trainer.use_native_amp:
+                # don't forget to retain graph on backward with multiple optimizers
+                self.trainer.scaler.scale(loss).backward()
```

Comment:
The `enabled=True|False` arguments to `GradScaler` and `autocast` help you integrate native amp without branches, ie, instead of
```
if self.trainer.use_native_amp:
    self.trainer.scaler.scale(loss).backward()
...
else:
    loss.backward()
```
you can construct `scaler` via `GradScaler(enabled=(use_native_amp and precision = 16))` then write
```
    self.trainer.scaler.scale(loss).backward()
```
to serve all cases (`scale` becomes a no-op if `scaler` is not enabled).

This applies more generally to the rest of the integration, but it's purely for convenience and since you've already have branches written, it's a matter of preference if you want to keep them.

Reply:
what about optimizers with closures?

Reply:
Closures aren't supported.  I haven't had anyone actually ask for one yet. https://github.com/PyTorchLightning/pytorch-lightning/pull/1561#discussion_r413237187
In principle I could implement closure handling entirely within, for example, `scaler.step(optimizer, closure=...)` but it's tricky.

Reply:
ok. we can throw an error for amp+second order optimizers not yet supported

Reply:
as you know, LBFGS is the only native optimizer that needs a closure, but other optimizers permit a closure.  I'm not sure if you provide the option for non-LBFGS optimizers to opt-into a closure, but if yes, you should include guidance that people must prefer non-closure usage for all optimizers if Amp is enabled.

Custom optimizers will also work as long as they apply gradients in a `step()` method (and that step method doesn't need a closure).

Reply:
got it. 
We'll keep it forked for now. Once we deprecate apex in 0.8.0 we'll do as you suggested!